cupyx/scipy/sparse/_index.py

```diff
@@ -0,0 +1,388 @@
+"""Indexing mixin for sparse matrix classes.
+"""
+
+import cupy
+from cupy import core
+
+import numpy
+
+
+_int_scalar_types = (int, numpy.integer)
+
+
+def _get_csr_submatrix(Ap, Aj, Ax,
+                       start_maj, stop_maj,
+                       start_min, stop_min):
+    """Return a submatrix of the input sparse matrix by
+    slicing both major and minor axes.
+
+    Args
+        Ap : indptr array from input sparse matrix
+        Aj : indices array from input sparse matrix
+        Ax : data array from input sparse matrix
+        start_maj : starting index of major axis
+        stop_maj : ending index of major axis
+        start_min : starting index of minor axis
+        stop_min : ending index of minor axis
+
+    Returns
+        Bp : indptr array of output sparse matrix
+        Bj : indices array of output sparse matrix
+        Bx : data array of output sparse matrix
+    """
+
+    Bp = Ap[start_maj:stop_maj+1]
+
+    start_offset = Bp[0]
+    stop_offset = Bp[-1]
+
+    Aj = Aj[start_offset:stop_offset]
+    Ax = Ax[start_offset:stop_offset]
+
+    Aj_copy = cupy.empty(Aj.size+1, dtype=Aj.dtype)
+    Aj_copy[1:] = Aj
+
+    Aj_copy[Aj_copy < start_min] = -1
+    Aj_copy[Aj_copy >= stop_min] = -1
+
+    mask = Aj_copy > -1
+
+    Aj_copy[mask] = 1
+    Aj_copy[~mask] = 0
+
+    cupy.cumsum(Aj_copy, out=Aj_copy)
+
+    Aj_copy -= Aj_copy[0]
+
+    Bp = Aj_copy[Bp-start_offset]
+    Bj = Aj[mask[1:]] - start_min
+    Bx = Ax[mask[1:]]
```

Comment:
Performance improvement. Could you confirm the performance numbers?

- This PR
```
- name = major_scalar, major = 10, minor = None, format = csr, density = 0.5, dtype = float32, n_rows = 15000, n_cols = 15000
cupy.sparse         :    CPU:  670.963 us   +/- 6.175 (min:  659.961 / max:  718.080) us     GPU-0:  677.899 us   +/- 6.243 (min:  666.624 / max:  725.312) us
- name = major_scalar, major = 10, minor = None, format = csc, density = 0.5, dtype = float32, n_rows = 15000, n_cols = 15000
cupy.sparse         :    CPU:39980.837 us   +/-11.082 (min:39956.465 / max:39999.470) us     GPU-0:41326.635 us   +/-10.980 (min:41302.177 / max:41344.383) us
- name = major_slice_minor_slice, major = slice(1, 500, None), minor = slice(1, 500, None), format = csr, density = 0.5, dtype = float32, n_rows = 15000, n_cols = 15000
cupy.sparse         :    CPU: 1766.492 us   +/- 5.824 (min: 1753.211 / max: 1788.307) us     GPU-0: 1788.933 us   +/- 5.741 (min: 1775.712 / max: 1810.080) us
- name = major_slice_minor_slice, major = slice(1, 500, None), minor = slice(1, 500, None), format = csc, density = 0.5, dtype = float32, n_rows = 15000, n_cols = 15000
cupy.sparse         :    CPU: 1772.303 us   +/-85.359 (min: 1751.207 / max: 3774.453) us     GPU-0: 1794.253 us   +/-85.011 (min: 1773.248 / max: 3788.224) us
- name = major_slice_minor_all, major = slice(1, 500, None), minor = slice(None, None, None), format = csr, density = 0.5, dtype = float32, n_rows = 15000, n_cols = 15000
cupy.sparse         :    CPU: 1672.464 us   +/- 5.794 (min: 1654.945 / max: 1694.381) us     GPU-0: 1713.061 us   +/- 5.628 (min: 1695.488 / max: 1735.168) us
- name = major_slice_minor_all, major = slice(1, 500, None), minor = slice(None, None, None), format = csc, density = 0.5, dtype = float32, n_rows = 15000, n_cols = 15000
cupy.sparse         :    CPU:39977.068 us   +/-64.564 (min:39931.267 / max:40278.025) us     GPU-0:41319.281 us   +/-47.469 (min:41277.218 / max:41530.750) us
- name = major_slice_minor_all, major = slice(500, 1, None), minor = slice(None, None, None), format = csr, density = 0.5, dtype = float32, n_rows = 15000, n_cols = 15000
cupy.sparse         :    CPU:  467.987 us   +/- 5.047 (min:  461.236 / max:  502.937) us     GPU-0:  475.003 us   +/- 5.128 (min:  468.384 / max:  510.560) us
- name = major_slice_minor_all, major = slice(500, 1, None), minor = slice(None, None, None), format = csc, density = 0.5, dtype = float32, n_rows = 15000, n_cols = 15000
cupy.sparse         :    CPU:37228.931 us   +/-11.117 (min:37203.334 / max:37241.228) us     GPU-0:37236.546 us   +/-11.181 (min:37210.815 / max:37248.672) us
- name = major_all_minor_slice, major = slice(None, None, None), minor = slice(5, 10, None), format = csr, density = 0.5, dtype = float32, n_rows = 15000, n_cols = 15000
cupy.sparse         :    CPU:39982.415 us   +/- 9.344 (min:39964.259 / max:40001.921) us     GPU-0:41329.272 us   +/- 9.325 (min:41310.944 / max:41348.385) us
- name = major_all_minor_slice, major = slice(None, None, None), minor = slice(5, 10, None), format = csc, density = 0.5, dtype = float32, n_rows = 15000, n_cols = 15000
cupy.sparse         :    CPU:  671.338 us   +/-57.881 (min:  660.658 / max: 2865.572) us     GPU-0:  678.335 us   +/-58.057 (min:  667.392 / max: 2879.040) us
```
- My suggestion
```
- name = major_scalar, major = 10, minor = None, format = csr, density = 0.5, dtype = float32, n_rows = 15000, n_cols = 15000
cupy.sparse         :    CPU:  383.086 us   +/-35.805 (min:  372.849 / max: 2107.814) us     GPU-0:  390.235 us   +/-35.966 (min:  379.808 / max: 2121.120) us
- name = major_scalar, major = 10, minor = None, format = csc, density = 0.5, dtype = float32, n_rows = 15000, n_cols = 15000
cupy.sparse         :    CPU:17638.530 us   +/- 8.464 (min:17619.170 / max:17655.693) us     GPU-0:18985.313 us   +/- 8.451 (min:18965.025 / max:19002.399) us
- name = major_slice_minor_slice, major = slice(1, 500, None), minor = slice(1, 500, None), format = csr, density = 0.5, dtype = float32, n_rows = 15000, n_cols = 15000
cupy.sparse         :    CPU:  814.139 us   +/- 4.864 (min:  803.106 / max:  837.460) us     GPU-0:  837.411 us   +/- 4.785 (min:  826.688 / max:  860.512) us
- name = major_slice_minor_slice, major = slice(1, 500, None), minor = slice(1, 500, None), format = csc, density = 0.5, dtype = float32, n_rows = 15000, n_cols = 15000
cupy.sparse         :    CPU:  814.042 us   +/- 4.589 (min:  802.338 / max:  833.884) us     GPU-0:  837.301 us   +/- 4.522 (min:  825.888 / max:  856.736) us
- name = major_slice_minor_all, major = slice(1, 500, None), minor = slice(None, None, None), format = csr, density = 0.5, dtype = float32, n_rows = 15000, n_cols = 15000
cupy.sparse         :    CPU:  877.460 us   +/- 4.375 (min:  866.231 / max:  901.897) us     GPU-0:  917.258 us   +/- 4.272 (min:  906.336 / max:  940.896) us
- name = major_slice_minor_all, major = slice(1, 500, None), minor = slice(None, None, None), format = csc, density = 0.5, dtype = float32, n_rows = 15000, n_cols = 15000
cupy.sparse         :    CPU:17684.164 us   +/-10.346 (min:17661.080 / max:17703.511) us     GPU-0:19030.526 us   +/-10.354 (min:19006.784 / max:19050.016) us
- name = major_slice_minor_all, major = slice(500, 1, None), minor = slice(None, None, None), format = csr, density = 0.5, dtype = float32, n_rows = 15000, n_cols = 15000
cupy.sparse         :    CPU:  195.288 us   +/- 6.910 (min:  190.776 / max:  316.687) us     GPU-0:  202.273 us   +/- 7.153 (min:  197.280 / max:  325.184) us
- name = major_slice_minor_all, major = slice(500, 1, None), minor = slice(None, None, None), format = csc, density = 0.5, dtype = float32, n_rows = 15000, n_cols = 15000
cupy.sparse         :    CPU:16261.189 us   +/- 8.451 (min:16234.323 / max:16281.635) us     GPU-0:16268.815 us   +/- 8.260 (min:16241.632 / max:16288.736) us
- name = major_all_minor_slice, major = slice(None, None, None), minor = slice(5, 10, None), format = csr, density = 0.5, dtype = float32, n_rows = 15000, n_cols = 15000
cupy.sparse         :    CPU:17638.589 us   +/- 9.062 (min:17614.987 / max:17661.698) us     GPU-0:18985.299 us   +/- 8.980 (min:18962.463 / max:19008.480) us
- name = major_all_minor_slice, major = slice(None, None, None), minor = slice(5, 10, None), format = csc, density = 0.5, dtype = float32, n_rows = 15000, n_cols = 15000
cupy.sparse         :    CPU:  382.056 us   +/- 4.320 (min:  375.580 / max:  406.016) us     GPU-0:  389.135 us   +/- 4.377 (min:  382.464 / max:  413.088) us
```

```suggestion
    # Major slicing
    Ap = Ap[start_maj : stop_maj + 1]
    start_offset, stop_offset = int(Ap[0]), int(Ap[-1])
    Ap = Ap - start_offset
    Aj = Aj[start_offset : stop_offset]
    Ax = Ax[start_offset : stop_offset]

    # Minor slicing
    mask = (start_min <= Aj) & (Aj < stop_min)
    Aj_copy = cupy.empty(Aj.size + 1, dtype=Aj.dtype)
    Aj_copy[0] = 0
    Aj_copy[1:] = mask
    cupy.cumsum(Aj_copy, out=Aj_copy)
    Bp = Aj_copy[Ap]
    Bj = Aj[mask] - start_min
    Bx = Ax[mask]
```